### PR TITLE
Refactor to TypeScript, several improvements, amm.

### PR DESCRIPTION
**Pull request by janbiasi**
> **DON'T MERGE - This is a work in progress PR, discussion will be needed regarding the feature set and this PR will provide the discussion base**

Background for this PR can be found [here](https://github.com/directus/sdk-js/issues/69#issuecomment-485003723).

## ToDos
- [x] Setup [Travis](https://travis-ci.org) for testing the SDK in multiple node versions. I'd suggest to run tests for the versions `8`, `9`, `10` (current LTS),`11` and `12` (LTS in october). As Node version 6 & 7 changed to maintenance LTS we should't support this further on. More about the [Node LTS plan](https://nodejs.org/de/about/releases/) on their website. We could include the badges later on to have a visual state of the SDKs stability status.
- [x] Include typings for all responses, must be validated against the Docs and example Directus system *(must)*
- [x] Fix the tests which are already broken in the current master branch (yes, there are several failing tests in the stable version), I'll come back to the owner later on. *(should)*
- [x] Strip the babel loader from the webpack configs as it's not needed anymore
- [x] Remove the dependency `node` as the context for looking inside the current dir's `node_modules/.bin` works absolutely fine. And at least it would be a `devDependency` rather than a `dependency`. (side note: it's also not supported on Node version 7)
- [x] Decouple the main HTTP/Request and Authentication stuff into own classes to obtain maintainability of the main core functionality of the SDK itself.
- [x] Refactor the `BodyType` definition to match the API spec.
- [x] Refactor the whole `substr(9)` / `directus_` prefix of collection handling to an own helper to eliminate the 10x duplication of code.
- [x] Clarify the question with an owner if the `env` key for in the SDKs constructor is still supported, it looks like that there are tests for this but no implementation for the setting.
- [x] Add a kinda `persit` flag to the config for controlling the de-/hydration of the configuration
- [x] ~~Refactor the signature of `request` for an easier usage and not line-long parameter list~~ Will create [a new Issue](#) for this one as it could involve breaking changes to the public API ...

## Ideas
- [x] ~Due heavy usage of generics in this update branch I'd suggest to switch from `spaces:2` to `tabs:4` to engage the readability of the codebase for other developers which will also lead to a better maintainability.~ **will be discussed on main directus repo**
- [x] Maybe strip the `argument-validator` out of the package and create something by our own, which is more simple and just do what we need to do, need to be clarified by myself first if this would be a good idea!

#### Restructure:
- decouple the SDK and the configuration, authentication, common HTTP layers into separate classes for clearer code.
  - use the `Configuration` class to manage the config, including re-/hydration with the optional provided storage
  - use the `Authentication` class to handle the token, login/-out mechanism and expiration
  - use the `API` class for common HTTP interactions such as get/put/post/delete/patch or the basic request function
- Global defaults can be set by importing `Configuration` from the package and setting `Configuration.defaults`. This is handy for configurations accross all SDK instances eg. `tokenExpirationTime` or multi-env SDKs for the same project.

#### Documentation:
- added a link to the recipes (see below) and state official TypeScript support so that developers don't need to search for issues to see if TS is supported officially.
- added recipes for popular frameworks to show how to use the SDK together with them (current recipes: Angular, React, RXJS, Redux, NGRX)

#### Tests:
- test files are renamed from `.ts` to `.spec.ts` to follow the general guidelines of all test frameworks to show visually that these are test files (even if they're located in a test folder!).
- broken tests are marked, search for the comment `FIXME: [ERR_STABLE]` or `TODO: [ERR_STABLE]` to find current broken tests on stable version.
- this tests need to be fixed, they don't work in the current stable release (tested it)
- updated **chai**, **sinon** and **mocha** to the latest versions
- integrated the `chai-datetime` package for the previous contained `.equalDate(...)` check in the `payload.spec.ts` file.
- Integrated travis CI for node versions 8, 9, 10, 11 and 12 (LTS in october)

#### Webpack:
- split configs to inherit from each other (better maintainability)
- add bundle analyzer for analyzing the bundle size

#### NPM:
- updated all dependencies (as of 12th of July 2019)
- all dependencies are now declared as fixed version (*)
- added all `@types` packages for typing, including tests
- save-exact added as default via `.npmrc` configuration (*)

> *)  If you're building out a production app, you don't always want to let your dependencies automatically update -- even if you've only specified a PATCH update, with hundreds of inter-working packages there is a possibility that an untested update could blow up your app. In development and when messing around with an app, you're probably safe allowing PATCH or MINOR updates. I believe when we are modeling a real-world scenario where you are building out an app that's going into production and you KNOW that the modules and all their dependencies work at the current (EXACT version) they are at now. 

#### NodeJS:
- remove support for node v7.x.x as maintenance ended this year
- bump to node v12.x.x (LTS starts in october)
- add nvm and node-version config file for version managers (`.nvmrc` and `.node-version`)

#### TypeScript:
- fixed all previous wrong typings and/or typing issues
- add typescript configurations for tests, source and declaration
- ~~integrate tslint for source- and test files (deprecated)~~
- integrate eslint for source- and test files (we should have a general convention for all files in an OpenSource project)
- compile declaration separately to dist/types to speed up the main build time
- all tests are now completely typed

> Refer to past issues about this request/problems via #37 and #69 

#### Updated prettier/editorconfig settings:
- differentiate between JSON, markdown and source files in the `.editorconfig`
- add additional config for single quotes and printWidth because it is the most common way in the JS world (see [bytearcher.com](https://bytearcher.com/articles/single-or-double-quotes-strings-javascript/) for refs to e.g. `async` or other popular libraries or the [popular and heavily used Airbnb](https://github.com/airbnb/javascript#strings) style guideline why we should do this)
- fix prettier script to include all JSONs, test- and source files (in endless levels, not just one layer deep)
- added the `quote_type = double` setting

> It is important to have a strictly defined code style for the whole project, not just the source code. 

#### Reduced the size of the main bundle:
- saved 8kb due direct `qs.stringify` import (parser not needed)

> It makes a pretty big difference whether you import a module like `import qs from 'qs'` or `import stringify from 'qs/lib/stringify'`, with the first one you'll include the whole package in the compiled output - which we don't need in our case.

### CODE REVIEW

**Comment by rijkvanzanten** [comment]
Great work 🙂 The only thing I'd argue is the code style changes (eg double quote / single quote, tabs vs spaces / etc). I know the reasoning behind it, but I'd really like the SDK to follow the style of the Directus app and all other Directus org projects. That being said, I don't mind moving the discussion there to optimize the style across the board!

**Comment by janbiasi** [comment]
@rijkvanzanten Thank you! I already thought that this will come 🙈 what would you propose? Leave it as it is or would you like to move the discussion regarding the code style?

**Comment by benhaynes** [comment]
I'd say let's open a discussion ticket on `directus/directus` regarding any updates to the code style — since that would need to be a global change. For now, we should keep everything consistent.

**Comment by rijkvanzanten** [comment]
@janbiasi I'm down to tweak the code style! I've already upped the line width from the default 80 to a 100 the other day ^^. I'm even willing to reconsider my nonsensical 2-spaces believes 😁 

The only thing that matters to me is that it's consistent between all Directus projects. Lets move those code style changes to the directus repo and implement it there first, then we can migrate all the other projects to use the same code style 🙂 

PS we're most likely not going to use TypeScript on the main application. I know that it brings a lot of niceties in it's type checking, but the added complexity in syntax makes it harder for new contributors to get started contributing. This is less of an issue with the SDK, as it's less prone to changes and fixes from the community. The main goal in terms of libraries / code style in the directus application itself is that it's as easy as possible to get started working on.

**Comment by janbiasi** [comment]
@rijkvanzanten I think it doesn't matter if you'd use TS on the main application because third party dev's won't access the main application code. Not so for the SDK, in times where nearly every web project uses TypeScript (speaking of sights from Namics, Switzerlands leading web agency) it's a requirement for a SDK to support it. 😄 I'll try to figure out all todos to discuss it, that I can finish the whole PR until next week.

**Comment by janbiasi** [comment]
Hey @rijkvanzanten, I just saw some configurations in test files which specify a `env` key while instanciating the SDK - but I found no reference to this in any code. My first thoughts were that this has been replaced by the `project`. Is it still needed or deprecated? 😸

**Comment by rijkvanzanten** [comment]
Good catch! `env` has been renamed to `project` to match the terminology used in the Directus API.

Thanks for taking the time to perfect the SDK. Helps out a lot ❤️

**Comment by janbiasi** [comment]
@rijkvanzanten 

That's why I love TypeScript 😂 so I've got a last few questions before finalizing all typings for the responses:

- [Returns the user preferences if there saved user preferences](https://github.com/directus/sdk-js/pull/72/files#diff-7d3b60bfa8a8e156bd5dc8838d807074R73), but it returns `collection` instead of `user`.
- [Returns the group preferences if there are no saved user preferences](https://github.com/directus/sdk-js/pull/72/files#diff-7d3b60bfa8a8e156bd5dc8838d807074R141), but it returns `collection` instead of `group`

The only one which works is the "Returns the collection preferences if there are no saved user or preferences" test. The issue was that in the test the filters for the mock response were declared as `filter[group][null]` instead of `filter[role][null]` (which the SDK calls).

I'm not familiar with the listing preferences API so it would be great to get a little bit support via e.g. a gist or something similar 😸

File: package.json [hunk 1]
```diff
@@ -25,35 +35,53 @@
...
   "bugs": {
-    "url": "https://github.com/directus/directus-sdk-javascript/issues"
+    "url": "https://github.com/directus/directus-sdk-js/issues"
   },
   "homepage": "https://docs.directus.io/sdk/js.html",
   "dependencies": {
-    "argument-validator": "^0.1.0",
-    "axios": "^0.18.0",
-    "base-64": "^0.1.0",
-    "node": "^11.12.0",
```
Comment (sarneeh) [inline package.json]: Maybe someone who added this wanted to ensure that it is being used with this particular `node` version range? How about replacing it with [the engines property](https://docs.npmjs.com/files/package.json#engines) which covers this use case?

Comment (janbiasi) [reply package.json]: Quite good idea to add the engines property, I've added a `.node-version` file for local development to specify the current runtime environment. Let me do this asap!

Tests are currently running for 7, 8, 9, 10 and 11. So I'd say that we should set the node engine to `>=7`, what do you think?

Comment (sarneeh) [reply package.json]: If there are tests which work with these particular versions, I think it's a good idea to set it just like you proposed :smiley:

**Comment by rijkvanzanten** [comment]
Hey @janbiasi! What's the current state of this? What else needs to be done?

**Comment by janbiasi** [comment]
Hey @rijkvanzanten, I've posted (& mentioned you) a few questions on 3rd of may which are still unanswered. The current state (except the things listed above) is:

- All tests are now up and running
- CI is integrated (travis)
- TSDoc is integrated to generate API docs for TypeScript reference (which you don't want to do by hand each time)
- Each method in the SDK supports now generic type arguments for typing responses for fetching collections and collection data
- Everything is typesafe
- The tokenExpiryTime feature was integrated

What needs to be done:
- Review by you and if possible by one or two other contributor(s)/owner(s)
- Version bump for breaking changes (major)

Could you start with a review/quality-control etc. that we can finalize everything and merge it?

**Comment by rijkvanzanten** [comment]
Thanks for the update 👌🏻 I'll get back to the other q's asap

> What I still don't get is why the compiled files are also published on GitHub as well as NPM, what's the case for this?

This is for people who don't rely on any buildchain or bundler.

**Comment by rijkvanzanten** [comment]
Re: extension endpoints: it returns an array of objects that is based on the meta.json files of the interfaces/layouts/pages respectively.

Re: failing tests
I don't know, but it sounds about right based on your description.

On what needs to be done:

> Review by you and if possible by one or two other contributor(s)/owner(s)

I'm not familiar with TypeScript, and I'm the only one who's working on the JS SDK. I have no choice but to trust your judgement. If you have a contact who's familiar with TypeScript and wants to review it that would be great @janbiasi. 

> Version bump for breaking changes (major)

What is the breaking change exactly? If it compiles down to the same build structure and it has the same methods etc etc, there shouldn't be a breaking change right?


One last thing that will be crucially important is having some docs on how to update all of this. Without this, it'll be a matter of weeks before someone (or me) opens a PR that doesn't update the typescript definitions correctly in turn defeating the purpose of this refactor in the first place..... 

Would be you down to co-own the SDK @janbiasi? I'm more than happy to invite you in as admin of the repo.

**Comment by janbiasi** [comment]
Hey @rijkvanzanten 

Before answering your questions a few things came into my mind which I have to check before I can give my 100% confidence with the PR:

1) Check all readme links (found two broken links)
2) Make a test integration into an existing project to find possible issues in regards of TS implementation

As I go to vacation this weekend I won't have time before 9th of July but I think I should be able to ensure everything until 14th of July (if that would be okay for you)

----

> Re: extension endpoints: it returns an array of objects that is based on the meta.json files of the interfaces/layouts/pages respectively.

Allright, I can't find a correct schema on the API docs for the types of the response, although it is generic at the moment so you can define which return type you want to get so it isn't a blocking thing atm. because with the types it just return `RequestPromise` which results in `Promise<any>` which then results in `any` so it's useless at the moment and you can just use `getLayouts()` which will return - without typings - just `Promise<any[]>` but you could override it like `getLayouts<MyLayoutType[]>()` and you'll receive your needed type as response.   

> Re: failing tests: I don't know, but it sounds about right based on your description.

Allright so I guess we'll leave that as it is, because the coverage increased dramatically and I think we can leave the test which fails atm. because it doesn't work. I will come back to it at a later time because I'm currently a bit under pressure job-wise and it doesn't hurry (imo)

> Review by you and if possible by one or two other contributor(s)/owner(s)
> I'm not familiar with TypeScript, and I'm the only one who's working on the JS SDK. I have no choice but to trust your judgement. If you have a contact who's familiar with TypeScript and wants to review it that would be great.

Okay that sounds good, I think I'll find one from my company who could take a brief look at it.

> Version bump for breaking changes (major): What is the breaking change exactly? If it compiles down to the same build structure and it has the same methods etc etc, there shouldn't be a breaking change right?

That's my fault, I messed something up I think, because the main API is still the same, except your newly introduced feature `tokenExpiryTime` which I renamed to `tokenExpirationTime` because it's grammatically the correct form (noun) instead of a substantive.

> One last thing that will be crucially important is having some docs on how to update all of this. Without this, it'll be a matter of weeks before someone (or me) opens a PR that doesn't update the typescript definitions correctly in turn defeating the purpose of this refactor in the first place.....

Completely get that, one of the benefits is that you literally can't break anything because if you build your PR and it fails because of wrong types you'll get the hint where somethings wrong. But I think we should add a basic/common `CONTRIBUTE.md` with some guidelines and help information about the environment, typings, basic concepts etc.

> Would be you down to co-own the SDK @janbiasi? I'm more than happy to invite you in as admin of the repo.

With pleasure! ❤️

**Comment by rijkvanzanten** [comment]
Thanks for the info!

> As I go to vacation this weekend I won't have time before 9th of July but I think I should be able to ensure everything until 14th of July (if that would be okay for you)

I'll be gone from next tuesday 'til the 9th of July too. Lets plan to rendezvous on [Slack](https://directus.chat) after the 9th when your back and comfortable about the current state of this PR. We can discuss how to handle the JS SDK moving forward as well 😄 

I'm very stoked about this ❤️

**Pull request merged**